lib/rack/throttle/rules.rb

```diff
@@ -27,29 +27,33 @@ def default_limit
       end
 
       def ips
-        @ips ||= options[:ip_whitelist].map { |ip| IPAddr.new(ip) } || []
+        @ips ||= (options[:ip_whitelist] || []).map { |ip| IPAddr.new(ip) } || []
       end
 
       def whitelisted?(request)
         return true if ip_whitelisted?(IPAddr.new(ip(request)))
-        return true if path_whitelisted?(request)
+        return true if rule_whitelisted?(request)
         false
       end
 
       def ip_whitelisted?(request_ip)
         !!ips.find { |ip| ip.include?(request_ip) }
       end 
 
-      def path_whitelisted?(request)
+      def rule_whitelisted?(request)
         rule = rule_for(request)
         rule ? rule[:whitelisted] : false
       end
 
       def rule_for(request)
-        rules.find do |rule|
-          next unless rule[:method] == request.request_method.to_s
-          next unless path_matches?(rule, request.path.to_s)
-          rule
+        cache.fetch("rule-#{request.object_id}") do 
```

Comment:
i added the caching as rule_for is called quite often and it is O(n) each time. There is no need to run through it that many times. Another option would be to rewrite the whole rack-throttle, to determine the cause once and pass it through the methods... as a cache is available, and a object-id should be unique somehow,.. maybe i should add an expire here after a few seconds..